encoder/libav_encoder.cpp

```diff
@@ -0,0 +1,548 @@
+{
+	const std::string codec_name("h264_v4l2m2m");
+
+	const AVCodec *codec = avcodec_find_encoder_by_name(codec_name.c_str());
+	if (!codec)
+		throw std::runtime_error("libav: cannot find video encoder " + codec_name);
+
+	codec_ctx_[Video] = avcodec_alloc_context3(codec);
+	if (!codec_ctx_[Video])
+		throw std::runtime_error("libav: Cannot allocate video context");
+
+	codec_ctx_[Video]->width = info.width;
+	codec_ctx_[Video]->height = info.height;
+	// usec timebase
+	codec_ctx_[Video]->time_base = { 1, 1000 * 1000 };
+	codec_ctx_[Video]->framerate = { (int)(options->framerate * 1000), 1000 };
+	codec_ctx_[Video]->pix_fmt = AV_PIX_FMT_DRM_PRIME;
+	codec_ctx_[Video]->sw_pix_fmt = AV_PIX_FMT_YUV420P;
+
+	if (info.colour_space)
+	{
+		using libcamera::ColorSpace;
+
+		static const std::map<ColorSpace::Primaries, AVColorPrimaries> pri_map = {
+			{ ColorSpace::Primaries::Raw, AVCOL_PRI_UNSPECIFIED },
+			{ ColorSpace::Primaries::Smpte170m, AVCOL_PRI_SMPTE170M },
+			{ ColorSpace::Primaries::Rec709, AVCOL_PRI_BT709 },
+			{ ColorSpace::Primaries::Rec2020, AVCOL_PRI_BT2020 },
+		};
+
+		static const std::map<ColorSpace::TransferFunction, AVColorTransferCharacteristic> tf_map = {
+			{ ColorSpace::TransferFunction::Linear, AVCOL_TRC_LINEAR },
+			{ ColorSpace::TransferFunction::Srgb, AVCOL_TRC_IEC61966_2_1 },
+			{ ColorSpace::TransferFunction::Rec709, AVCOL_TRC_BT709 },
+		};
+
+		static const std::map<ColorSpace::YcbcrEncoding, AVColorSpace> cs_map = {
+			{ ColorSpace::YcbcrEncoding::None, AVCOL_SPC_UNSPECIFIED },
+			{ ColorSpace::YcbcrEncoding::Rec601, AVCOL_SPC_SMPTE170M },
+			{ ColorSpace::YcbcrEncoding::Rec709, AVCOL_SPC_BT709 },
+			{ ColorSpace::YcbcrEncoding::Rec2020, AVCOL_SPC_BT2020_CL },
+		};
+
+		auto it_p = pri_map.find(info.colour_space->primaries);
+		if (it_p != pri_map.end())
+			codec_ctx_[Video]->color_primaries = it_p->second;
+
+		auto it_tf = tf_map.find(info.colour_space->transferFunction);
+		if (it_tf != tf_map.end())
+			codec_ctx_[Video]->color_trc = it_tf->second;
+
+		auto it_cs = cs_map.find(info.colour_space->ycbcrEncoding);
+		if (it_cs != cs_map.end())
+			codec_ctx_[Video]->colorspace = it_cs->second;
+
+		codec_ctx_[Video]->color_range =
+			info.colour_space->range == ColorSpace::Range::Full ? AVCOL_RANGE_JPEG : AVCOL_RANGE_MPEG;
+	}
+
+	// Apply any options->
+	if (options->bitrate)
+		codec_ctx_[Video]->bit_rate = options->bitrate;
+
+	if (!options->profile.empty())
+	{
+		static const std::map<std::string, int> profile_map = {
+			{ "baseline", FF_PROFILE_H264_BASELINE },
+			{ "main", FF_PROFILE_H264_MAIN },
+			{ "high", FF_PROFILE_H264_HIGH }
+		};
+
+		auto it = profile_map.find(options->profile);
+		if (it == profile_map.end())
+			throw std::runtime_error("libav: no such profile " + options->profile);
+
+		codec_ctx_[Video]->profile = it->second;
+	}
+
+	if (!options->level.empty())
+	{
+	}
+
+	if (options->inline_headers)
+	{
+	}
```

Comment:
Interesting, so there's no option for repeating the headers in every I frame. I guess we don't care? I suppose if you were streaming elementary H.264 to a socket then you'd have no reason to crank up the whole of FFmpeg, and otherwise it would take care of that kind of thing (headers quite often go out-of-band after all, don't they).

Reply:
Yes, this is more a limitation of the libav interface to the v4lm2m codec (as far as I can tell).  But luckily libav does manage to do the right thing internally when outputting to files/containers vs streaming.

Reply:
At the V4L2 level it does set V4L2_CID_MPEG_VIDEO_HEADER_MODE to HEADER_MODE_SEPARATE to get the headers in a separate buffer to the I-frame, but it never sets V4L2_CID_MPEG_VIDEO_REPEAT_SEQ_HEADER.

It could repeat the headers within the FFmpeg code, but it doesn't ask V4L2 to do it.